codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameWriter.java

```diff
@@ -498,15 +593,70 @@ private ChannelFuture writeHeadersInternal(ChannelHandlerContext ctx,
 
             // Encode the entire header block.
             headerBlock = ctx.alloc().buffer();
-            headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            try {
+                headersEncoder.encodeHeaders(streamId, headers, headerBlock);
+            } catch (Http2Exception e) {
+                return promise.setFailure(e);
+            }
 
             Http2Flags flags =
                     new Http2Flags().endOfStream(endStream).priorityPresent(hasPriority).paddingPresent(padding > 0);
-
-            // Read the first fragment (possibly everything).
             int nonFragmentBytes = padding + flags.getNumPriorityBytes();
             int maxFragmentLength = maxFrameSize - nonFragmentBytes;
-            ByteBuf fragment = headerBlock.readRetainedSlice(min(headerBlock.readableBytes(), maxFragmentLength));
+            int fragmentBytes = min(headerBlock.readableBytes(), maxFragmentLength);
+            // if the first fragment fully consume the headerBlock, we can try making it fit into the
+            // header frame if is small enough: it would avoid using a promise aggregator and reduce the number
+            // of pipeline traversal
+            if (fragmentBytes <= COPY_SLICE_CUTOFF_BYTES && fragmentBytes == headerBlock.readableBytes() &&
+                paddingBytes(padding) <= 64) {
+                // no need for a continuation either!
+                try {
+                    int paddingBytes = paddingBytes(padding);
+                    flags.endOfHeaders(true);
+                    final int bufBytes =
+                            HEADERS_FRAME_HEADER_LENGTH + fragmentBytes + (paddingBytes > 0? paddingBytes : 0);
+                    ByteBuf buf = ctx.alloc().buffer(bufBytes);
+                    int payloadLength = fragmentBytes + nonFragmentBytes;
+                    writeFrameHeaderInternal(buf, payloadLength, HEADERS, flags, streamId);
+                    writePaddingLength(buf, padding);
+                    if (hasPriority) {
+                        buf.writeInt(exclusive? (int) (0x80000000L | streamDependency) : streamDependency);
+                        // Adjust the weight so that it fits into a single byte on the wire.
+                        buf.writeByte(weight - 1);
+                    }
+                    buf.writeBytes(headerBlock, fragmentBytes);
+                    if (paddingBytes > 0) {
+                        // this is quite fast assuming small padding
+                        // see https://github.com/netty/netty/pull/13693
+                        buf.writeZero(paddingBytes);
+                    }
+                    ctx.write(buf, promise);
+                } catch (Throwable t) {
+                    promise.setFailure(t);
+                    PlatformDependent.throwException(t);
```

Comment:
why re throw ? feels wrong

Reply:
Goog point; I've blindly copied from the original code this one iirc, but as you said, doesn't look right 👍

Reply:
I don't feel yet comfortable to change it yet, i see that throwing the exception allow to correctly release in the caller the expected "leaky" data. I would leave it as it is, unless you want me to address the original code behaviour now (which is fine, just need to know you opinion here)